Accept transforms in gutenberg_get_global_styles context params

samnajian opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
This PR addresses the changes explained in https://github.com/WordPress/gutenberg/issues/49712 by introducing the `transforms` to the `$context` param as form of an array, so that 
``` 
gutenberg_get_global_styles( array(), array( 
    'block_name' => 'core/post-terms',
    'transforms' => array( 'resolve-variables' ) 
    )
 );
```
returns:

```
"core/post-terms": {
    "typography": { "fontSize": "12px" }
}

```
and not the following:
```
"core/post-terms": {
    "typography": { "fontSize": "var(--wp--preset--font-size--small)" }
}
```

for a `theme.json` with the following content:

```
"core/post-terms": {
    "typography": { "fontSize": "var(--wp--preset--font-size--small)" }
}
```

## Why?
There are some usages of the `gutenberg_get_global_styles`  where the user is interested in the values of the css rules and not the variables.

## How?

Currently the acceptable value is only `resolve-variables` as in 
```
$context = array( 'transforms' => array( 'resolve-variables' ) ) 
```
and in future PRs other transforms can be added such as 
```
$context =  array( 'transforms' => array( 'resolve-references' ) )
```

## Testing Instructions
Paste the following in `functions.php` of the theme:

```
add_action( 'init', function(){
        error_log( print_r( wp_get_global_styles( array(), array('block_name'=>'core/post-terms') ), true ) );
} );
```

samnajian committed:
Define WP_Theme_JSON_Gutenberg::get_style_values

The function is defined to return styles with values
instead of variables

samnajian commented on lib/class-wp-theme-json-gutenberg.php:
```diff
@@ -3578,4 +3578,49 @@ protected function get_feature_declarations_for_node( $metadata, &$node ) {
+		$preset_vars = static::compute_preset_vars( self::get_settings(), static::VALID_ORIGINS );
+		$theme_vars  = static::compute_theme_vars( self::get_settings() );
+		$vars        = array_reduce(
+			array_merge( $preset_vars, $theme_vars ),
```
@oandregal I'd like to know your idea on this, would this produce any unseen problem?

oandregal replied:
Presets and custom properties defined by the theme and the only source of data we have available, so this is fine.

oandregal commented on lib/compat/wordpress-6.3/get-global-styles-and-settings.php:
```diff
@@ -158,3 +158,38 @@ function gutenberg_get_global_styles( $path = array(), $context = array() ) {
+ * @see wp_get_global_styles()
+ * @return array The styles to retrieve with their values.
+ */
+function gutenberg_get_global_style_values( $path = array(), $context = array() ) {
```
How do you envision this function to be used with the other one (resolving the references)? In other words, what's the API to compose these transformations?

Through the conversation we've had, I see two options forward:

- **Standalone functions**. We'd have `gutenberg_transform_style_from_variable_to_value` and `gutenberg_transform_style_from_reference_to_value`, respectively. We can compose them:

```php
gutenberg_get_global_styles(); // Some consumers don't want those transformations.
gutenberg_transform_style_from_variable_to_value( gutenberg_get_global_styles() ); // Some do not want vars.
gutenberg_transform_style_from_reference_to_value( gutenberg_get_global_styles() ); // Some do not want refs.
gutenberg_transform_style_from_variable_to_value( gutenberg_transform_style_from_reference_to_value( gutenberg_get_global_styles() ) ); // Some do not want either.
```

The new functions should be idempotent, meaning that `get styles > transform from variables > transform from references` should yield the same result as `get styles > transform from references > transform from variables`.

- **Parameter to the existing function**. Either expanding the existing `$context` or a new `$transforms` parameter. Not super opinionated about it.

```php
gutenberg_get_global_styles(); // Some consumers don't want the transformations.
gutenberg_get_global_styles( array(), array( 'transforms' => array( 'resolve-variables' ) ) ); // Some do not want vars.
gutenberg_get_global_styles( array(), array( 'transforms' => array( 'resolve-references' ) ) ); // Some do not want refs.
gutenberg_get_global_styles( array(), array( 'transforms' => array( 'resolve-references', 'resolve-variables' ) ) ); // Some do not want either.
```

I'm leaning towards the parameter approach.

oandregal replied:
Pinging some people that had been involved in these conversations at https://github.com/WordPress/gutenberg/issues/49712 https://github.com/WordPress/gutenberg/issues/49715 and https://github.com/WordPress/gutenberg/issues/45171

cc @jsnajdr @ramonjd @geriux @Aljullu @jeherve @BogdanUngureanu 

Is there any other alternative? What are your thoughts?

ramonjd replied:
I'm more attracted to the idea of controlling the output via parameters for the reasons you mentioned — some users might want ref resolution only etc.

The function signature already features two arguments: `gutenberg_get_global_styles( $path = array(), $context = array() ) ` — are we thinking of loading the `$context` with transform options? I think that sounds reasonable, just making sure ;) Backwards compat should be unaffected.

Looking at the code in this PR and even the [messy function I wrote to resolve node refs](https://github.com/WordPress/gutenberg/pull/49737/files#diff-b03597cc3da199e5aa5a94950e8241135904853e7c3b82d758e42744878afae7R3049), it doesn't look like it will bloat the function too much.

samnajian replied:
@oandregal Initially I was more leaned towards a new function, but after adding that and figuring that technically the function parameters stay the same, I'm also more inclined towards a parameter approach, and thanks to your suggested example, I see more clearly that this approach is more viable.

samnajian committed:
Make gutenberg_get_global_styles accept transforms in context param

Once transforms => array( 'resolve_variables' ) is defined the css
variables are resolved to their values

oandregal commented on phpunit/class-wp-theme-json-test.php:
```diff
@@ -2136,4 +2136,111 @@ public function test_internal_syntax_is_converted_to_css_variables() {
+						),
+					),
+				),
+				'styles'   => array(
```
Thanks for adding tests for so many cases. Can I ask for also adding a test for top-level styles? It'd be great to also have some data that uses raw values and not custom properties, to make sure those work fine as well.

oandregal commented on phpunit/class-wp-theme-json-test.php:
```diff
@@ -2136,4 +2136,111 @@ public function test_internal_syntax_is_converted_to_css_variables() {
+			)
+		);
+
+			$styles = $theme_json->get_styles_with_values();
```
My editor (VSCode) reports that something is off with the indentation. I opened the file with the `less` command to make sure it wasn't my editor, and I still see how the indentation is off.

![image](https://github.com/WordPress/gutenberg/assets/583546/50089f93-f720-4600-9408-831e7d6d2163)

I wonder why the lint job is not reporting it :thinking: I also run `npm run format:php` (which is supposed to fix this kind of things) and it didn't find any fixable errors.

oandregal replied:
I'm not sure what happens, so I'm going to summon people who know more than I do on this topic and have worked in this area recently: @ObliviousHarmony @noahtallen @anton-vlasenko Would you know why this linting issue is not reported in CI? I cannot get it to fail locally, either.

I changed some of the tabs to spaces, and it fails (as it should). Then I created this piece of art (which is a linting issue) and `npm run lint:php` still passes:

![image](https://github.com/WordPress/gutenberg/assets/583546/63a59507-d224-43b7-a956-300f2bdfc67a)

oandregal replied:
@samnajian in the meantime, would you mind fixing this manually?

oandregal replied:
Actually, `npm run lint:php` fails due to some permissions:

![image](https://github.com/WordPress/gutenberg/assets/583546/1ee5c86f-11ff-46a6-af00-dc1b7d07307b)

noahtallen replied:
I couldn't replicate the permissions issue, though it could probably be resolved by re-creating the container with wp-env (we recently updated its permissions model to ideally work better in most cases).

However, you can execute `composer run-script lint` locally without needing wp-env. (If you have the composer binary installed yourself, that is.) On this branch, that doesn't give me any errors either. Given that CI isn't reporting errors either, there must be a difference between Gutenberg's config and whatever your editor is using... it seems like Gutenberg's phpcs config may not be handling indentation properly? That's outside my realm of expertise though 😅

samnajian replied:
@oandregal should be fixed in 0781b38438ee4cde66dc53e6dec3f16c9e7b20e6

oandregal commented on lib/global-styles-and-settings.php:
```diff
@@ -261,8 +266,16 @@ function gutenberg_get_global_styles( $path = array(), $context = array() ) {
+	&& is_array( $context['transforms'] )
+	&& in_array( 'resolve-variables', $context['transforms'], true );
+
+	if ( $resolve_variables ) {
```
This works, but the idea is to have more transforms available. How would they be composable when we have two of them? Note https://github.com/WordPress/gutenberg/pull/50484#discussion_r1192502874 that we can have 4 use cases:

1. no transforms
2. only `resolve-variables`
3. only `resolve-references`
4. both `resolve-variables` and `resolve-references`

I worry that if we create a method in the `WP_Theme_JSON_Gutenberg` class for each of the use cases from 2-4 implementing the 4th becomes specially complex to read.

What if we make the transformations composable and standalone operations? For example, we could have one function for transformation: 

```php
/**
 * @access private 
 * @param array $styles The style nodes to transform.
 * @return The transformed style nodes.
 */
function _gutenberg_transform_theme_json_styles_resolve_variables( $styles ) {}
```

This way, the code at `wp_get_global_styles` becomes much simpler as we evolve it:

```php
$styles = WP_Theme_JSON_Resolver_Gutenberg::get_merged_data( $origin )->get_raw_data()['styles'];

if ( $resolve_variables ) {
  $styles = _gutenberg_transform_theme_json_styles_resolve_variables( $styles );
}

if ( $resolve_references ) {
  $styles = _gutenberg_transform_theme_json_styles_resolve_references( $styles );
}

return $styles;
```

What do you think?

samnajian replied:
If I understand you correctly, you are suggesting to move the logic for these transformation from private class methods (i.e. `WP_Theme_JSON_Gutenberg::get_styles_with_values`) out of the `WP_Theme_JSON_Gutenberg` and into a function like `_gutenberg_transform_theme_json_styles_resolve_variables` but annotate it as private? And in both cases we should resolve the value for `$resolve_variables` and later `$resolve_references` inside `wp_get_global_styles`, right?

It'll prevent the class from growing bigger for sure, but I'm not getting how it will make it more readable.
The only advantage I see in your suggestion is having **composable** functions like `_gutenberg_transform_theme_json_styles_resolve_variables`, while `WP_Theme_JSON_Gutenberg::get_styles_with_values` is by nature static, and we can make it static and allow passing the `$styles` from outside, like the following:

```
public static function get_styles_with_values( $styles = null ) {
     ....
     $_styles = $styles && is_array($styles) ? $styles : self::get_raw_data()['styles'];
      ....
     return self::convert_variables_to_value( $styles, $vars );
}
```

That said I'm fine either way.

oandregal replied:
Yeah, either a function or a public static method of the private class `WP_Theme_JSON_Gutenberg` is fine by me. Perhaps I'm leaning toward separate functions because of the size of the class, though it's fine to start with the static method. Not opinionated about that.

What I care about is that the the transformation is a pure operation that takes `$styles` as input and returns the transformed `$styles` as output: `$styles = transformation( $styles )`.

oandregal replied:
Looking at the existing code, I have a suggestion to make, what if we do this?

```php
// ...
$merged_data = WP_Theme_JSON_Resolver_Gutenberg::get_merged_data( $origin );
if ( $resolve_variables ) {
    $merged_data = WP_Theme_JSON_Gutenberg::resolve_variables( $merged_data );
}
$styles = $merged_data->get_raw_data()['styles'];
// ...
```

Notes:

- The code that executes the transformations could take a `$theme_json` as input and return the same structure. This is so consumers don't need to know the details of the inner operation (does it use settings? etc.?), simplifying the consumer code. It'd also make all transformations use the same API. Note this is what we already do for the existing `resolve_custom_css_format`, so building on that would be great.
- Ideally, the name of the method in the theme.json class is representative of the name of the parameter for the transformation. For example, if the parameter is called `resolve-variables`, I'd expect the method to be called something like `resolve_variables`.

samnajian replied:
addressed in fff3c43707e3e8f36ed7b9c15427b34a4a93f7ac

geriux commented:
Hey there! I just did a quick test of these changes to see what the response would be like and **this would benefit the mobile editor a lot** since we won't need to go through the data and manually parse it (which on mid-end devices could take a toll on performance).

I can't add too much value to the code review though since my PHP knowledge is limited 😅 but I like that it still uses `gutenberg_get_global_styles` and it just sets a context to get the style values. Once we implement this we would need to keep support for the default response so this works perfectly.

To test it I updated the mobile endpoint to pass the transforms:

```php
$context = array(
	'transforms' => array(
		'resolve-variables',
	),
);			
$settings['__experimentalStyles'] = gutenberg_get_global_styles(array(), $context);
```

Thank you for working on this @samnajian 👏

samnajian committed:
Add more test to assert raw style values stay the same

samnajian committed:
Convert WP_Theme_JSON_Gutenberg::get_styles_with_values to static resolve_style_values

This will allow a more composable interface to this class

oandregal commented on phpunit/class-wp-theme-json-test.php:
```diff
@@ -2136,4 +2136,121 @@ public function test_internal_syntax_is_converted_to_css_variables() {
+				'styles'   => array(
+					'color'    => array(
+						'background' => 'var(--wp--preset--color--primary)',
+						'text'       => $raw_color_value,
```
Could we add also a couple of test cases for the following values?

- `var(--undefined-css-property)`: custom css property is not defined by the presets. I presume the expected result is this case is that we let it unresolved.
- `calc(var(--wp--preset--font-size--small) + 20px);`: the custom css property is not necessarily the first thing we'll find in the value.

samnajian replied:
Great idea, I'd include cases for gradiant color as well as in `'linear-gradient(90deg, var(--wp--preset--color--primary) 0%, var(--wp--preset--color--secondary) 35%, rgba(0,212,255,1) 100%)'`

samnajian replied:
addressed in 9cf3d846b71922a53550f17f25377a66847a90a4

oandregal replied:
I found some other cases we need to support: fallback values for the variable. Examples:

- `var(--wp--preset--font-size--small, fallbackValue)` should be resolved to the `--wp--preset--font-size--small` value.
- `var(--wp--preset--font-size--small, var(--wp--preset--font-size--medium))` should be resolved to the `--wp--preset--font-size--small` value.
- `var(--undefined-property, var(--wp--preset--font-size--medium))` should be left as it is.

samnajian replied:
Addressed in ce6903883897f2eb7617a582d9a858d60c6a1876

ramonjd commented:
Do folks think we should flag this for inclusion in 6.3? Would it be ready and tested? (Code freeze in about 2 weeks)

Thank you! 🙇

oandregal commented:
> Do folks think we should flag this for inclusion in 6.3? Would it be ready and tested? (Code freeze in about 2 weeks)

I was hoping to merge this before the end of this week, so it'd be part of 16.1. According to the conversation in [the 6-3-release-leads slack channel](https://wordpress.slack.com/archives/C051Z1SKBDZ/p1686662805023139) it looks like 16.1 is the last gutenberg version bundled for 6.3? If so, this could make it. Otherwise, no need to cherry-pick.

samnajian committed:
Make sure CSS variables with var in all places get transformed correctly

ramonjd commented:
> looks like 16.1 is the last gutenberg version bundled for 6.3? If so, this could make it. Otherwise, no need to cherry-pick.

That's my understanding too. Thanks @oandregal !! 🍺 

I can add it to the [backport tracking issue](https://github.com/WordPress/gutenberg/issues/51077) once merged as well just to keep things all together.

oandregal commented on phpunit/class-wp-theme-json-test.php:
```diff
@@ -2136,4 +2136,135 @@ public function test_internal_syntax_is_converted_to_css_variables() {
 
 	}
+
+	public function test_get_global_styles_values() {
```
```suggestion
	public function test_resolve_variables() {
```

oandregal commented on lib/class-wp-theme-json-gutenberg.php:
```diff
@@ -3646,4 +3646,67 @@ private static function resolve_custom_css_format( $tree ) {
+	private static function convert_variables_to_value( $styles, $values ) {
+		foreach ( $styles as $key => $style ) {
+			if ( is_array( $style ) ) {
+				$styles[ $key ] = self::convert_variables_to_value( $style, $values );
```
We could explicitly add `continue` after this is processed, instead of letting the continuing happening later because it's not a string.

oandregal commented on lib/class-wp-theme-json-gutenberg.php:
```diff
@@ -3646,4 +3646,67 @@ private static function resolve_custom_css_format( $tree ) {
+			}
+
+			if ( 0 === strpos( $style, 'var(' ) ) {
+				$styles[ $key ] = isset( $values[ $style ] ) ? $values[ $style ] : $style;
```
This is assuming that the value will come as `var(--wp--preset--font-size--small)` but it could also have fallback values as in `var(--wp--preset--font-size--small, 12px)`. I mentioned this in the test thread as a case to cover https://github.com/WordPress/gutenberg/pull/50484/files#r1229447871

samnajian committed:
Add support for more cases and variations

oandregal commented on phpunit/class-wp-theme-json-test.php:
```diff
@@ -2021,4 +2021,152 @@ public function test_internal_syntax_is_converted_to_css_variables() {
+		$this->assertEquals( $large_font, $styles['blocks']['core/navigation']['elements']['link']['typography']['fontSize'], 'Block element: font-size' );
+
+		$this->assertEquals(
+			'var(--undefined--color--primary, var(--wp--preset--font-size--secondary))',
```
Shouldn't we resolve the fallback value `var(--wp--preset--font-size--secondary)`?

oandregal replied:
Ah, I see how my comment at https://github.com/WordPress/gutenberg/pull/50484/files#r1229447871 may have been confusing: 

> var(--undefined-property, var(--wp--preset--font-size--medium)) should be left as it is.

To clarify, I think it'd be best to leave undefined as they are, but always resolve the ones we can. The expected result here would be `var(--undefined-property, $resolved_value)`.

What do you think?

oandregal reviewed (commented):
This is brilliant work, Sam! Thanks for all your effort here. There's [only one use case](https://github.com/WordPress/gutenberg/pull/50484/files#r1232446298) we need to fix, but otherwise it's working as expected.

samnajian committed:
Resolve fallback variable in case the main variable is undefined

samnajian committed:
Make sure WP_Theme_JSON_Gutenberg::resolve_variables returns $theme_json

oandregal reviewed (approved):
This is ready to be merged. Thanks for pushing until it was feature complete, Sam! It was harder than anticipated, and you did great.

I see some lint issues that need fixing before merging, though.

ramonjd commented:
Thanks everyone for working on this. I'll add it to the [backport list](https://github.com/WordPress/gutenberg/issues/51077) for WP 6.3 🙇

oandregal commented:
Backporting at https://github.com/WordPress/wordpress-develop/pull/4656
